### PR TITLE
Add test to dataset name cache

### PR DESCRIPTION
**Pull request by evalott100**
Adds a test to dataset name cache.

Also
* pins numpy until https://github.com/mdavidsaver/p4p/issues/145 is solved.
* adds some docstrings and comments.
* changes `:CAPTURE` to a `ComboBox` widget in the positions capture table.

### CODE REVIEW

File: src/pandablocks_ioc/_hdf_ioc.py [hunk 1]
```diff
@@ -332,10 +337,11 @@ def hdf_writer_names(self):
         """Formats the current dataset names for use in the HDFWriter"""
 
         hdf_names: Dict[str, Dict[str, str]] = {}
-        for record_name, dataset in self.datasets.items():
+        for record_name, dataset in self._datasets.items():
             if not dataset.name or dataset.capture == "No":
                 continue
 
+            # COUNTER:OUT -> COUNTER.OUT
             field_name = record_name.replace(":", ".")
```
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_hdf_ioc.py:345]: Shouldn't this be using the `epics_to_panda_name` conversion function?

File: src/pandablocks_ioc/_hdf_ioc.py [hunk 2]
```diff
@@ -332,10 +337,11 @@ def hdf_writer_names(self):
         """Formats the current dataset names for use in the HDFWriter"""
 
         hdf_names: Dict[str, Dict[str, str]] = {}
-        for record_name, dataset in self.datasets.items():
+        for record_name, dataset in self._datasets.items():
```
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_hdf_ioc.py:340]: It looks like the typing of the `_datasets` is not entirely accurate? If I understand its usage correctly, it's more accurately typed as a `Dict[EpicsName, Dataset]`?

File: pyproject.toml [hunk 3]
```diff
@@ -13,7 +13,7 @@ classifiers = [
 description = "Create an IOC from a PandA"
 dependencies = [
     "setuptools>=64",
-    "numpy",
+    "numpy<2",
```
Comment (AlexanderWells-diamond) [inline pyproject.toml:16]: Probably worth adding the p4p issue number here too, to explain why it's pinned.

File: src/pandablocks_ioc/_hdf_ioc.py [hunk 4]
```diff
@@ -313,13 +313,18 @@ def handle_data(self, data: HDFReceived):
 
 @dataclass
 class Dataset:
+    """A dataset name and capture mode"""
```
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_hdf_ioc.py:316]: What is a "Dataset" in this context? I don't know enough about this area to know if this is a commonly known topic, or a concept created here for our code.

**Review by AlexanderWells-diamond — APPROVED** [review]
Looks fine but there's a couple of possible improvements that could be made

File: src/pandablocks_ioc/_hdf_ioc.py (hunk 2 above)
Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:340]: That's correct. I'll change.

File: src/pandablocks_ioc/_hdf_ioc.py (hunk 4 above)
Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:316]: It's more of an abstraction - we have the `BLOCK_NAME:DATASET` record for using an alternative column label in the hdf5 file and the `BLOCK_NAME:CAPTURE` record for specifying what is captured. `class Dataset` functions as a more readable way to keep these together in the dict given from `ioc.py`.

**Pull request merged**